_posts/2019-09-19-ijwb-reverse-engineering-sync.md

```diff
@@ -0,0 +1,513 @@
+projects.
+
+```protobuf
+message BlazeProjectData {
+  reserved 1;
+  TargetMap target_map = 2 [deprecated = true];
+  BlazeInfo blaze_info = 3;
+  BlazeVersionData blaze_version_data = 4;
+  WorkspacePathResolver workspace_path_resolver = 5;
+  WorkspaceLanguageSettings workspace_language_settings = 6;
+  SyncState sync_state = 7;
+  TargetData target_data = 8;
+}
+```
+
+Most of the language specific data are in `TargetData`'s `TargetIdeInfo`,
+defined in
+[`intellij_ide_info.proto`](https://github.com/bazelbuild/intellij/blob/master/proto/intellij_ide_info.proto).
+This is where the
+[aspect](https://blog.bazel.build/2016/06/10/ide-support.html)-generated
+`intellij-info.txt` files come into play:
+
+```protobuf
+message TargetIdeInfo {
+  string kind_string = 1;
+  TargetKey key = 2;
+  ArtifactLocation build_file_artifact_location = 3;
+  repeated Dependency deps = 4;
+  repeated string tags = 5;
+  repeated string features = 6;
+  TestInfo test_info = 7;
+
+  // The time this target was most recently synced, in milliseconds since epoch.
+  // Not provided by the aspect directly; instead filled in by the plugin during
+  // sync.
+  int64 sync_time_millis = 20;
+
+  JavaIdeInfo java_ide_info = 100;
+  JavaToolchainIdeInfo java_toolchain_ide_info = 101;
+
+  AndroidIdeInfo android_ide_info = 110;
+  AndroidAarIdeInfo android_aar_ide_info = 111;
+  AndroidSdkIdeInfo android_sdk_ide_info = 112;
+  AndroidInstrumentationInfo android_instrumentation_info = 113;
+
+  CIdeInfo c_ide_info = 120;
+  CToolchainIdeInfo c_toolchain_ide_info = 121;
+
+  PyIdeInfo py_ide_info = 130;
+  GoIdeInfo go_ide_info = 140;
+  JsIdeInfo js_ide_info = 150;
+  TsIdeInfo ts_ide_info = 160;
+  DartIdeInfo dart_ide_info = 170;
+  KotlinToolchainIdeInfo kt_toolchain_ide_info = 180;
+}
+```
+
+`TargetIdeInfo` is heart of the plugin's multi-language support. During a sync,
+each Bazel target is associated with an `intellij-info.txt` file that contains a
+text representation of the `TargetIdeInfo` proto. Here's an example for the main
+Spring Boot `java_binary` target:
+
+```prototxt
+build_file_artifact_location {
+  is_external: false
+  is_new_external_version: true
+  is_source: true
+  relative_path: "src/main/java/hello/BUILD.bazel"
+  root_execution_path_fragment: ""
+}
+deps {
+  dependency_type: 0
+  target {
+    label: "@local_config_cc//:toolchain"
+  }
+}
+deps {
+  dependency_type: 0
+  target {
+    label: "@bazel_tools//tools/jdk:current_java_toolchain"
+  }
+}
+deps {
+  dependency_type: 1
+  target {
+    label: "//src/main/java/hello:lib"
+  }
+}
+java_ide_info {
+  jars {
+    jar {
+      is_external: false
+      is_new_external_version: true
+      is_source: false
+      relative_path: "src/main/java/hello/app.jar"
+      root_execution_path_fragment: "bazel-out/darwin-fastbuild/bin"
+    }
+    source_jar {
+      is_external: false
+      is_new_external_version: true
+      is_source: false
```

Comment:
how is this false? Also below? When is this true? Maybe it's not important to this post, your call

Reply:
Thanks. Adding something for 1).

2-5 is out of scope for my intention of this post. The working set is already [documented](https://ij.bazel.build/docs/bazel-plugin.html#expand-sync-to-working-set), and `derive_targets_from_directories` is a relatively new feature for project management that is not directly impacting how the sync process works internally.

Reply:
`is_source` is true only for artifacts that are source code to your project. e.g. genrule outputs are not `is_source`. It has nothing to do with source jars.